### ISSUE TITLE
Recovery middleware

### ISSUE THREAD

**Issue by pjebs** [urfave/negroni#134]
I noticed that relatively recently the recovery middleware got added: 
`ErrorHandlerFunc func(interface{})`.

If `rec.ErrorHandlerFunc(err)` panics, then recovery middleware will be useless because it is _itself_ inside the `recover()`

**Comment by pjebs** [comment]
On second thoughts, I guess you can add 2 recovery middleware one after another.

**Comment by jszwedko** [comment]
@pjebs you are correct in that -- it would be the passer's responsibility to catch any `panic`s thrown within their handler. They could do so with something like:

```go
rec := negroni.NewRecovery()
rec.ErrorHandlerFunc = func(foo interface) {
  defer func() {
    if err := recover(); err != nil {
      fmt.Println("panic'd in our recovery")
    }
  }()

  // something that might panic
}
```

**Comment by pjebs** [comment]
Seems a bit dangerous to me. If we expected the user to have to make sure their code was bullet proof, we wouldn't even need the recovery middleware. The recovery middleware should be the final wall of protection.

**Comment by jszwedko** [comment]
@pjebs I agree, which is the default behavior. If you are specifying a custom handler, I'd argue that you take on the responsibility of writing code that shouldn't itself `panic` (similar to handling trapping signals in other languages)., but I concede that this could be better documented.

The alternative is that we wrap the call to user's function in another anonymous function with its own `defer` to catch any possible `panic`s. I'd be OK with this as long as it was possible to opt-out of the behavior. I'll think on that a bit more.

**Comment by pjebs** [comment]
I think the best option is what you stated: wrap with its own defer.
I can't image *ANY* reason to be able to opt out.

**Comment by jszwedko** [comment]
It could conceivably be the case that you never expect your handler to fail and would prefer that the app die (and perhaps be restarted by a process manager), but I admit that it is unlikely.

**Comment by pjebs** [comment]
well why would they use recovery middleware then?

**Comment by jszwedko** [comment]
@pjebs I mean if their handler `panic`d, not if the original request caused a `panic`.

**Comment by pjebs** [comment]
Let them call `os.Exit()` from inside their handler _if_ that is their desire, perhaps inside the error handler's defer.

**Comment by jszwedko** [comment]
Sure, either way we introduce work for the party that wants the opposite behavior. I think you are correct that it the more _expected_ behavior is to recover the additional `panic`s for the user though.

**Comment by jszwedko** [comment]
Put together #138 with a possible approach

**Comment by jszwedko** [comment]
Closed by #138 -- thanks for bringing this up @pjebs !
